In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
from sklearn.metrics import calinski_harabasz_score

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [5]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()
scaled_market_data = scaler.fit_transform(df_market_data[["price_change_percentage_24h","price_change_percentage_7d",
                                                   "price_change_percentage_14d","price_change_percentage_30d",
                                                   "price_change_percentage_60d","price_change_percentage_200d",
                                                   "price_change_percentage_1y"]])

In [6]:
# Create a DataFrame with the scaled data
df_scaled_market_data = pd.DataFrame(scaled_market_data, columns=["price_change_percentage_24h","price_change_percentage_7d",
                                                   "price_change_percentage_14d","price_change_percentage_30d",
                                                   "price_change_percentage_60d","price_change_percentage_200d",
                                                   "price_change_percentage_1y"])

# Copy the crypto names from the original DataFrame
df_scaled_market_data["coin_id"] = df_market_data.index

# Set the coin_id column as index
df_scaled_market_data.set_index("coin_id", inplace=True)


# Display the scaled DataFrame
df_scaled_market_data

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [7]:
# Create a list with the number of k-values from 1 to 11
market_k_values = list(range(2, 12)) 

In [8]:
# Create an empty list to store the inertia values
market_inertia_values = []
market_score_values = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in market_k_values:
    k_model = KMeans(n_clusters=i, random_state=1)
    labels = k_model.fit_predict(df_scaled_market_data)
    labels = k_model.labels_
    score = calinski_harabasz_score(df_scaled_market_data, labels)
    print(i, score)
    market_inertia_values.append(k_model.inertia_)
    market_score_values.append(score)

2 13.766470160788646
3 14.021116879878372
4 32.45985294262567
5 29.892958092798118
6 25.59559866274402
7 29.631263481193304
8 26.466078342876497
9 30.8795987518146
10 29.000071118892503
11 29.514981261832265


In [9]:
# Create a dictionary for k-value scores
k_scores_market = {"k": market_k_values, "score": market_score_values}

# Create a DataFrame for the scores and display in descending order
k_scores_market_df = pd.DataFrame(k_scores_market)
k_scores_market_df.sort_values("score", ascending=False)

k      score
2   4  32.459853
7   9  30.879599
3   5  29.892958
5   7  29.631263
9  11  29.514981
8  10  29.000071
6   8  26.466078
4   6  25.595599
1   3  14.021117
0   2  13.766470

In [10]:
# Create a dictionary with the data to plot the Elbow curve
market_elbow_curve = {"k": market_k_values, "inertia": market_inertia_values}

# Create a DataFrame with the data to plot the Elbow curve
df_market_elbow_curve = pd.DataFrame(market_elbow_curve)
df_market_elbow_curve.sort_values("inertia", ascending=False)

k     inertia
0   2  212.123342
1   3  165.136752
2   4   79.022435
3   5   66.413051
4   6   61.634088
5   7   46.074467
6   8   43.392694
7   9   32.913223
8  10   30.469111
9  11   26.480101

In [11]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
df_market_elbow_curve.hvplot.line(
    x="k", 
    y="inertia", 
    title="Optimal K Value for Scaled Market Data", 
    xticks=market_k_values
)
 

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** The best value for K is **4** as it occurs at the elbow point of the curve. 


---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [12]:
# Initialize the K-Means model using the best value for k
best_market_kvalue = 4  
model = KMeans(n_clusters=best_market_kvalue, random_state=1)

In [13]:
# Fit the K-Means model using the scaled DataFrame
model.fit(df_scaled_market_data)

KMeans(n_clusters=4, random_state=1)

In [14]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
predictions = model.predict(df_scaled_market_data)
df_scaled_market_data['cluster'] = predictions
# Print the resulting array of cluster values.
print(predictions)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 3 2 0 0 1
 0 0 0 0]


In [15]:
# Create a copy of the scaled DataFrame
df_scaled_market_data_copy = df_scaled_market_data.copy()

In [16]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
df_scaled_market_data_copy['cluster'] = predictions
# Display the copy of the scaled DataFrame
df_scaled_market_data_copy

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [17]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
# Create a scatter plot using hvPlot
scatter_plot_market = df_scaled_market_data_copy.hvplot.scatter(
    x="price_change_percentage_24h", 
    y="price_change_percentage_7d",   
    by="cluster",                       
    hover_cols=["coin_id"],       
    title="Cryptocurrency Changes by Market Components", 
    xlabel="24h Price Change (%)",     
    ylabel="7d Price Change (%)"       
)

# Display the plot
scatter_plot_market

:NdOverlay   [cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [18]:
# Create a PCA model instance and set `n_components=3`.
pca_model = PCA(n_components=3)

In [19]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
pca_transformed = pca_model.fit_transform(df_scaled_market_data)

# View the scaled PCA data
pca_transformed

array([[ 4.48907952e-01,  1.24537573e+00,  8.50640396e-01],
       [ 4.95367093e-01,  8.99822995e-01,  1.31755943e+00],
       [-8.18845712e-01, -7.18990907e-02, -6.95015071e-01],
       [-8.40357106e-01, -8.00540676e-02, -5.44359519e-01],
       [ 8.13240044e-01,  2.66952035e+00,  1.64332113e+00],
       [ 8.22870558e-01,  1.68209607e+00,  9.05203458e-01],
       [ 9.13867626e-01,  1.12868393e+00,  2.73003754e+00],
       [ 8.11102222e-01,  1.11337793e+00,  1.59865515e+00],
       [ 2.58728083e-01,  9.46883804e-01,  1.10355990e+00],
       [-1.00392082e+00,  1.19410202e-01, -2.78563993e-01],
       [-1.36328853e+00, -1.24534877e+00, -9.50712158e-01],
       [-8.30643324e-01, -7.70471599e-02, -7.07660520e-01],
       [-1.13545778e+00, -1.71400623e-01, -4.41569215e-01],
       [ 2.08497797e+00,  2.68897925e+00,  7.76469068e-01],
       [-6.33198072e-01,  1.69113551e-02, -2.49426354e-01],
       [-1.61907992e-01,  3.10904366e-01,  1.81852107e+00],
       [-9.96986646e-01, -1.29257396e+00

In [20]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
explained_variance = pca_model.explained_variance_ratio_
# Print the explained variance for each principal component
print("Explained Variance for Each Principal Component:")
for i, variance in enumerate(explained_variance, start=1):
    print(f"PC{i}: {variance:.4f}")

Explained Variance for Each Principal Component:
PC1: 0.3487
PC2: 0.3136
PC3: 0.2263


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 
- PC1: 0.3487
- PC2: 0.3136
- PC3: 0.2263

In [21]:
# Create a new DataFrame with the PCA data.
pca_transformed_df = pd.DataFrame(pca_transformed, columns=['PC1', 'PC2', 'PC3'])

# Copy the crypto names from the original DataFrame
pca_transformed_df["coin_id"] = df_market_data.index

# Set the coin_id column as index
pca_transformed_df.set_index("coin_id", inplace=True)

# Display the scaled PCA DataFrame
pca_transformed_df.head()

PC1       PC2       PC3
coin_id                                   
bitcoin       0.448908  1.245376  0.850640
ethereum      0.495367  0.899823  1.317559
tether       -0.818846 -0.071899 -0.695015
ripple       -0.840357 -0.080054 -0.544360
bitcoin-cash  0.813240  2.669520  1.643321

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [22]:
# Create a list with the number of k-values from 1 to 11
pca_k_values = list(range(2, 12)) 

In [23]:
# Create an empty list to store the inertia values
pca_inertia_values = []
pca_score_values = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in pca_k_values:
    k_model = KMeans(n_clusters=i, random_state=1)
    labels = k_model.fit_predict(pca_transformed_df)
    labels = k_model.labels_
    score = calinski_harabasz_score(pca_transformed_df, labels)
    print(i, score)
    pca_inertia_values.append(k_model.inertia_)
    pca_score_values.append(score)

2 15.673798098263996
3 25.846649417614714
4 68.88800460849342
5 69.74192741660832
6 80.03246346396583
7 72.96847856891921
8 68.6816149059634
9 81.22590744451753
10 93.86470105341775
11 106.5562170923258


In [24]:
# Create a dictionary for k-value scores
pca_elbow_data = {"k": pca_k_values, "score": pca_score_values}

# Create a DataFrame for the scores and display in descending order
df_pca_scaled_data = pd.DataFrame(pca_elbow_data)
df_pca_scaled_data.sort_values("score", ascending=False)

k       score
9  11  106.556217
8  10   93.864701
7   9   81.225907
4   6   80.032463
5   7   72.968479
3   5   69.741927
2   4   68.888005
6   8   68.681615
1   3   25.846649
0   2   15.673798

In [25]:
# Create a dictionary with the data to plot the Elbow curve
pca_elbow_data = {"k": pca_k_values, "inertia": pca_inertia_values}

# Create a DataFrame with the data to plot the Elbow curve
pca_elbow_df = pd.DataFrame(pca_elbow_data)
pca_elbow_df

k     inertia
0   2  207.307439
1   3  123.126951
2   4   44.130602
3   5   33.217427
4   6   23.374713
5   7   20.943077
6   8   18.666961
7   9   13.640110
8  10   10.287149
9  11    7.958180

In [26]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
pca_elbow_df.hvplot.line(
    x="k", 
    y="inertia", 
    title="Optimal K Value for Scaled PCA Data", 
    xticks=pca_k_values
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **4**


* **Question:** Does it differ from the best k value found using the original data?

  * **No** 

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [27]:
# Initialize the K-Means model using the best value for k
pca_k_values = list(range(2, 12)) 

In [28]:
# Fit the K-Means model using the PCA data
model.fit(pca_transformed_df)

KMeans(n_clusters=4, random_state=1)

In [29]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
cluster_prediction = model.predict(pca_transformed_df)

# Print the resulting array of cluster values.
cluster_prediction

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 3, 1, 0, 0, 2, 0, 0, 0, 0])

In [30]:
# Create a copy of the scaled PCA DataFrame
pca_transformed_df_copy = pca_transformed_df.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
pca_transformed_df_copy['cluster'] = cluster_prediction

# Display the copy of the scaled PCA DataFrame
pca_transformed_df_copy

PC1       PC2       PC3  cluster
coin_id                                                    
bitcoin               0.448908  1.245376  0.850640        1
ethereum              0.495367  0.899823  1.317559        1
tether               -0.818846 -0.071899 -0.695015        0
ripple               -0.840357 -0.080054 -0.544360        0
bitcoin-cash          0.813240  2.669520  1.643321        1
binancecoin           0.822871  1.682096  0.905203        1
chainlink             0.913868  1.128684  2.730038        1
cardano               0.811102  1.113378  1.598655        1
litecoin              0.258728  0.946884  1.103560        1
bitcoin-cash-sv      -1.003921  0.119410 -0.278564        0
crypto-com-chain     -1.363289 -1.245349 -0.950712        0
usd-coin             -0.830643 -0.077047 -0.707661        0
eos                  -1.135458 -0.171401 -0.441569        0
monero                2.084978  2.688979  0.776469        1
tron                 -0.633198  0.016911 -0.249426        0
tezos                -0.161908  0.310904  1.818521        1
okb                  -0.996987 -1.292574 -0.793701        0
stellar              -0.990512 -0.449877 -0.052002        0
cosmos                0.878129  1.093136  1.774115        1
cdai                 -0.874815 -0.008933 -0.694993        0
neo                  -1.135199 -0.792113 -0.499484        0
wrapped-bitcoin       0.434801  1.232506  0.839794        1
leo-token            -1.110348 -0.540593 -0.854934        0
huobi-token          -0.758629 -0.094085 -0.466658        0
nem                   0.222636 -0.083977 -1.584956        0
binance-usd          -0.826142 -0.049070 -0.700753        0
iota                 -1.059844 -0.097044  0.149781        0
vechain              -1.808829 -1.520236 -0.527335        0
zcash                 0.558082  0.920479  1.377145        1
theta-token           1.452903 -1.756424 -2.506866        0
dash                 -0.897830 -0.134758  0.197015        0
ethereum-classic     -0.908899 -0.095612 -0.194364        0
ethlend               6.049082 -7.033268  2.235723        3
maker                 0.307206  0.565760  0.968488        1
havven               -0.507111 -2.407910  0.392697        0
omisego              -0.236488  0.071399 -1.445063        0
celsius-degree-token  6.454065  2.823519 -4.810646        2
ontology             -1.823588 -1.561754 -0.000663        0
ftx-token            -0.689158  0.253900 -0.073233        0
true-usd             -0.823298 -0.031016 -0.694746        0
digibyte             -0.770668 -0.187671 -0.911020        0

In [31]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
scatter_plot_pca = pca_transformed_df_copy.hvplot.scatter(
    x='PC1',
    y='PC2',
    by='cluster',  
    hover_cols=['coin_id'],  
    title='Cryptocurrency Changes by PCA Components'
)
scatter_plot_pca

:NdOverlay   [cluster]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [32]:
# Composite plot to contrast the Elbow curves
plt.figure(figsize=(12, 6))
# Display market elbow plot
market_elbow_plot = df_market_elbow_curve.hvplot.line(
    x="k", 
    y="inertia", 
    title="Optimal K Value for Scaled Market Data", 
    xticks=market_k_values
)
# Display PCA elbow plot
pca_elbow_plot = pca_elbow_df.hvplot.line(
    x="k", 
    y="inertia", 
    title="Optimal K Value for Scaled PCA Data", 
    xticks=pca_k_values
)
#Create composite elbow plot
composite_elbow_plot = market_elbow_plot + pca_elbow_plot 
#Display composite elbow plot
composite_elbow_plot

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

<Figure size 1200x600 with 0 Axes>

In [33]:
# Composite plot to contrast the clusters
plt.figure(figsize=(12, 6))
# Implement market scatter plot
scatter_plot_market = df_scaled_market_data_copy.hvplot.scatter(
    x="price_change_percentage_24h", 
    y="price_change_percentage_7d",   
    by="cluster",                       
    hover_cols=["coin_id"],       
    title="Cryptocurrency Changes by Market Components", 
    xlabel="24h Price Change (%)",     
    ylabel="7d Price Change (%)"       
)
#Implement PCA scatter plot
scatter_plot_pca = pca_transformed_df_copy.hvplot.scatter(
    x='PC1',
    y='PC2',
    by='cluster',  
    hover_cols=['coin_id'],  
    title='Cryptocurrency Changes by PCA Components'
)
#Create composite scatter plot
composite_cluster_plot = scatter_plot_market + scatter_plot_pca 
#Display composite scatter plot
composite_cluster_plot

:Layout
   .NdOverlay.I  :NdOverlay   [cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [cluster]
      :Scatter   [PC1]   (PC2,coin_id)

<Figure size 1200x600 with 0 Axes>

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

      * **Answer:** Using fewer features to cluster data with K-means leads to misleading data due to missing components that are crucial to a detailed analysis.  